plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java

```diff
@@ -171,16 +178,35 @@ public CloseableIterator<List<Object>> iterator()
             addCloseable(planFilesIterator);
 
             return new CloseableIterator<>() {
+                private Iterator<StructLike> contentFilesIterator = emptyIterator();
+
                 @Override
                 public boolean hasNext()
                 {
-                    return !closed && planFilesIterator.hasNext();
+                    if (!closed && contentFilesIterator.hasNext()) {
+                        return true;
+                    }
+
+                    if (!closed && (planFilesIterator.hasNext())) {
+                        DataTask dataTask = (DataTask) planFilesIterator.next();
+                        contentFilesIterator = dataTask.rows().iterator();
+                        return contentFilesIterator.hasNext();
+                    }
+
+                    return false;
                 }
 
                 @Override
                 public List<Object> next()
                 {
-                    return getRecord(planFilesIterator.next().file());
+                    StructLike dataTaskFile = contentFilesIterator.next();
```

Comment:
This is not quite bulletproof.

If you call `next()` several times in a row and you happen to reach the end of `contentFilesIterator`, there will be thrown a `NoSuchElementException` even though there are other file scan tasks to iterate.

Please build the iterator to be more robust.

Reply:
Have a look for a PoC into https://github.com/apache/iceberg/blob/3b37f581371c1fd70d2a56f5bd900ec7a8436e7e/flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java#L107-L136

Reply:
Oh...my initial thought is that `next()` will always come after `hasNext()` check, so the `NoSuchElementException` exception can be avoid. But I think your reference code is much more clear, let me build the iterator in a more robust way. Many thanks for your detailed reference!